### DISCUSSION TITLE
Async plan?

### DISCUSSION THREAD

**Discussion by osa1** [koka-lang/koka#620] [General]
I'm curious what the plan is for implementing async support in the current version of Koka? I found #384, but it was created a year ago and last updated 7 months ago. Is this on hold? Can we expect async libraries any time soon?

**Comment by TimWhiting** [comment]
Hi @osa1, I'm the author of that PR and am in touch with Daan. There are a few things with the type system & implicits, as well as better low level bytes / string primitives that it seems like he wants to do prior to finishing the LibUV integration. The branch should be still useable, but lacks HTTP/S support (only TCP + UDP / file support currently). 

Also of interest is probably Daan's response [here](https://github.com/koka-lang/koka/discussions/581#discussioncomment-10847665), in which he talks a bit more about his plans.

In the unofficial community libraries I have copied some of the JS backend part of the async support in the PR (which doesn't need any language / runtime changes). 
Here is an [example](https://github.com/koka-community/std/blob/main/examples/async/timer.kk) of how to use it.
Add the `-i../path/to/std` after cloning to use the community std library. 

That library also has more examples of how to use Koka's extern functions - C/JS ffi, which you can use to bring other async functions you are missing if you are willing to write some native code. (Which I'm assuming you are comfortable with seeing as you work on Dart). 

If you have more questions with regards to those repositories in general the Koka enthusiasts have been conversing on Discord:  https://discord.gg/CEkHHnTqEv

Daan has been busy recently with [mimalloc](https://github.com/microsoft/mimalloc) (a memory allocator) which he works on for microsoft research and which has recently been adopted by Python for memory allocation, so that has demanded a lot of his time, but we are meeting soon to discuss next priorities.

**Reply by osa1** [reply]
Thanks for the details, looking forward to seeing async support in the current version of the language.

**Reply by osa1** [reply]
> In the unofficial community libraries I have copied some of the JS backend part of the async support in the PR (which doesn't need any language / runtime changes).
Here is an [example](https://github.com/koka-community/std/blob/main/examples/async/timer.kk) of how to use it.
Add the -i../path/to/std after cloning to use the community std library.

What is the full command? I'm trying your example with:

```
koka --target=js test.kk -i../std
```

But it's printing a lot of warnings and then failing:

```
...
warning: backend does not support external in std/num/int64/@extern-clmul-wide
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int64/clmulr
warning: backend does not support external in std/num/int64/bit-zip
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int64/clmulr
warning: backend does not support external in std/num/int64/bit-zip
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int64/clmulr
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int64/check   : std/num/float64
clmulr
warning: backend does not support external in std/num/int64/clmul
warning: backend does not support external in std/num/int32/@extern-clmul_wide
check   : std/num/decimal
check   : std/num/ddouble
check   : std/time/date
check   : std/time/timestamp
check   : std/time/duration
check   : std/async/timer
/home/omer/Downloads/koka/share/koka/v3.1.2/lib/std/os/null.kk(36, 7): type error: identifier maybe cannot be resolved.
  context      :       maybe
  inferred type: _
  candidates   : null/maybe  : forall<a> (n : null<a>) -> maybe<a>
                 string/maybe: (s : string) -> maybe<string>
                 int/maybe   : (i : int) -> maybe<int>
                 either/maybe: forall<a,b> (e : either<a,b>) -> maybe<b>
                 maybe/maybe : forall<a,b,e> (m : maybe<a>, onNothing : b, onJust : (a) -> e b) -> e b
                 maybe       : forall<a> (t : error<a>) -> maybe<a>
  hint         : give a type annotation or qualify the name?

Failed to compile test.kk
```

**Reply by TimWhiting** [reply]
Ahh, I forgot there is an error in the std null library which is used by `async.kk` (it should work with the dev branch of Koka - or with a simple change to `null.kk`). 

Your command is correct. 

Go to `home/omer/Downloads/koka/share/koka/v3.1.2/lib/std/os/null.kk` and change the last definition to the following:
```koka
// Cast a boolean `False` to null
pub fun bool/null( b : bool ) : null<()>
    if b then Just(()).null else Nothing.null
```

Then recompile as you did before:
`koka --target=js test.kk -i../std`

And finally run the `.mjs` file using `node xx.mjs`

**Reply by TimWhiting** [reply]
BTW, I also just updated the `koka-community/std` repo with a temporary `std/async/null.kk` file with this change, and updated `std/async` to depend on that so you can use it without messing with the packaged std libraries. You'll just need to pull the latest commit.

**Reply by osa1** [reply]
Thanks, it works now.

**Reply by osa1** [reply]
@TimWhiting I'm trying to find where is the `await` function used here is defined: https://github.com/koka-community/std/blob/f98bee331da917569a28c0cfb331d40484632275/std/async/async.kk#L426-L430

The only `await` definition that I could find is here: https://github.com/koka-lang/koka/blob/21089c158ef19dbecf0a3d934a04b8ecbff20125/lib/std/os/task.kk#L41-L42

But it calls a C function: https://github.com/koka-lang/koka/blob/21089c158ef19dbecf0a3d934a04b8ecbff20125/lib/std/os/task.kk#L26-L27

So it can't be used in JS backend. How is `async` defined in the JS backend?

**Reply by TimWhiting** [reply]
It's a bit hidden by the comment above it, but it is just two definitions down: https://github.com/koka-community/std/blob/f98bee331da917569a28c0cfb331d40484632275/std/async/async.kk#L443.

To drill down to the native code / effect: 
- https://github.com/koka-community/std/blob/f98bee331da917569a28c0cfb331d40484632275/std/async/async.kk#L473
- https://github.com/koka-community/std/blob/f98bee331da917569a28c0cfb331d40484632275/std/async/async.kk#L574
- https://github.com/koka-community/std/blob/f98bee331da917569a28c0cfb331d40484632275/std/async/async.kk#L547
- Though the closest handler could also be a cancelable scope or interleaved async computations which mostly just forward the async handling to the main async handler
- https://github.com/koka-community/std/blob/f98bee331da917569a28c0cfb331d40484632275/std/async/async.kk#L380
- https://github.com/koka-community/std/blob/f98bee331da917569a28c0cfb331d40484632275/std/async/async.kk#L528

Note that there is not really a need for any native code to be involved except if a computation wants to yield or wait early. Those functions use implicits to require that the platform provide `set-timeout` equivalent functions to schedule a callback on the event loop. These are defined in https://github.com/koka-community/std/blob/main/std/async/timer.kk, and exposed along with async here: https://github.com/koka-community/std/blob/main/std/async.kk.

Also, note that I just took the async support from the version 1 std libraries and updated it to use the new keywords, etc. 
https://github.com/koka-lang/koka/tree/dev/lib/v1/std
